### DISCUSSION TITLE
Automatic tracing for AWS SDK for JavaScript

### DISCUSSION THREAD

**Discussion by dreamorosi** [aws-powertools/powertools-lambda-typescript#276] [Maintainers discussions]
**Problem statement**
The Python version of Powertools [automatically traces](https://awslabs.github.io/aws-lambda-powertools-python/latest/core/tracer/#patching-modules) for AWS SDK clients (and other modules). This is possible thanks to the fact that X-Ray SDK for Python supports checking which modules are imported.

On the other hand AWS currently offers two versions of the AWS SDK for JavaScript ([v2](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/welcome.html) & [v3](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/)) with only the former bundled in the Javascript runtime and with the latter implementing a [modular architecture](https://aws.amazon.com/blogs/developer/modular-packages-in-aws-sdk-for-javascript/).

This discussion aims at finding (& documenting) an agreement about the direction that this version of Powertools should take.

**Current behaviour**
`Tracer` allows Customers to opt-in into patching all clients using the method `captureAWS` (only for v2) or specific clients using the methods `captureAWSClient` and `captureAWSv3Client`.

These methods are lifted directly from the X-Ray SDK for JavaScript:
- `captureAWS` [API doc](https://docs.aws.amazon.com/xray-sdk-for-nodejs/latest/reference/module-aws_p.html#.captureAWS) - [source](https://github.com/aws/aws-xray-sdk-node/blob/master/packages/core/lib/patchers/aws_p.js#L30)
- `captureAWSClient` [API doc](https://docs.aws.amazon.com/xray-sdk-for-nodejs/latest/reference/module-aws_p.html#.captureAWSClient) - [source](https://github.com/aws/aws-xray-sdk-node/blob/master/packages/core/lib/patchers/aws_p.js#L55)
- `captureAWSv3Client` [source](https://github.com/aws/aws-xray-sdk-node/blob/master/packages/core/lib/patchers/aws3_p.ts#L186)

**Potential paths forward**

**Option 1**
Leave the decision to Customers and let them decide which client to patch. This would also be the easiest path forward as `Tracer` wouldn't have to know whether v2 and/or v3 is being used.

**Option 2**
Automatically patch all AWS SDK v2 clients by using the `captureAWS` method and relying on the fact that a version of v2 should always be present in the execution environment. Customers that are using v3 would still have to explicitly patch all clients.

**Comment by flochaz** [comment]
From my point of view, we should not make this decision and let the customer do it with  X-Ray SDK already providing an easy enough usable API.

**Reply by dreamorosi** [reply]
Agree with you but happy to hear also other opinions; if we all agree then let's go this way.

**Comment by heitorlessa** [comment]
As TS isn’t as clear cut as Python, I’d vote for option 1 — letting them
decide.

This will change either way when the X-Ray team officially makes an
announcement they told me today — I’ll share in tomorrow’s sync

**Comment by saragerion** [comment]
Conclusion: we are not going to do it.
